### PR TITLE
Goals: full feature parity with Expenses

### PR DESCRIPTION
**Pull request by douglasrlee**
## Summary

- Adds a full Goals CRUD page with the same Simple-style display as Expenses (Allocated/Total summary, per-paycheck breakdown, progress bars, off-track state, tap-to-edit drawer)
- Extends the allocation engine to fund goals from paycheck schedules alongside expenses
- Adds transaction-to-goal linking (via `GoalLinker`) mirroring the existing expense linker — transactions can be assigned to a fully-funded goal bucket
- Updates Free-to-Spend to subtract goal bucket balances alongside expense buckets

## Database changes

- New `goals` table (mirrors `expenses`: name, amount, cadence, due_on, funding_schedule_id, user_id)
- `allocations.expense_id` made nullable; new `allocations.goal_id` FK added; partial unique indexes replace the old blanket unique index; CHECK constraint enforces exactly one of expense_id/goal_id is set
- `transactions.goal_id` FK added for goal linking

## Test plan

- [ ] All 388 existing tests pass (`bin/rails test`)
- [ ] Navigate to `/goals` — empty state with + button
- [ ] Add a goal — appears in list with progress bar, Allocated/Total header, per-paycheck amount
- [ ] Edit a goal — drawer pre-fills, saves correctly
- [ ] Delete a goal — returns to empty state
- [ ] Open a transaction drawer — "Goal" section appears below "Expense" section; both are mutually exclusive (linking to one hides the picker for the other)
- [ ] Free-to-Spend on transactions page reflects funded goal buckets

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: app/javascript/controllers/goal_picker_controller.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { Controller } from "@hotwired/stimulus"
+
+export default class extends Controller {
+  static targets = ["input", "label", "option", "submit"]
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/javascript/controllers/goal_picker_controller.js:4]: Fixed in e09a5ba — `goal-picker` is now imported and registered in `app/javascript/controllers/index.js`. Good catch; the picker UI was dead without it.

File: db/migrate/20260622053816_create_goals_and_extend_allocations_for_goals.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    # Exactly one of expense_id or goal_id must be set
+    add_check_constraint :allocations,
+                         "num_nonnulls(expense_id, goal_id) = 1",
+                         name: "allocations_exactly_one_allocatable",
+                         validate: false
+
+    add_column :transactions, :goal_id, :uuid
+    add_index :transactions, :goal_id, algorithm: :concurrently
+    add_foreign_key :transactions, :goals, validate: false
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply db/migrate/20260622053816_create_goals_and_extend_allocations_for_goals.rb]: Fixed in e09a5ba — added `on_delete: :nullify` to the `transactions -> goals` FK. Without it, deleting a goal with a linked transaction raised an FK violation (neither `Goal` nor `Expense` declares `has_many :transactions`, so nullification relies entirely on the DB-level action). Mirrors the earlier transactions/expenses fix.

File: db/migrate/20260622053816_create_goals_and_extend_allocations_for_goals.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+              unique: true,
+              where: "goal_id IS NOT NULL",
+              name: "index_allocations_on_funding_event_id_and_goal_id",
+              algorithm: :concurrently
+
+    # Exactly one of expense_id or goal_id must be set
+    add_check_constraint :allocations,
+                         "num_nonnulls(expense_id, goal_id) = 1",
+                         name: "allocations_exactly_one_allocatable",
+                         validate: false
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply db/migrate/20260622053816_create_goals_and_extend_allocations_for_goals.rb:49]: Addressed in e09a5ba — added a follow-up migration `20260622053817_validate_goals_foreign_keys_and_constraints` that validates all four new FKs and the `allocations_exactly_one_allocatable` CHECK constraint, matching the existing `Validate*` pattern in the repo. Rolled back and re-migrated to confirm both run clean.

File: app/models/goal.rb [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  end
+
+  def per_paycheck_amount(as_of: Date.current)
+    next_due = next_due_on(after: as_of)
+    return amount unless next_due && funding_schedule
+
+    paychecks = funding_schedule.occurrence_count_between(after: as_of, through: next_due)
+    paychecks = 1 if paychecks.zero?
+    (amount / paychecks).round(2)
+  end
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/models/goal.rb:48]: Added in e09a5ba — `test/models/goal_test.rb` mirrors `expense_test.rb`: validations, `funding_schedule_belongs_to_user`, cadence advance/recede + day-clamping edge cases, `next_due_on`, `bucket_balance`, `off_track?`, `fully_funded?`, and `per_paycheck_amount` (including past-due roll-forward).

File: app/controllers/goals_controller.rb [hunk 5]
```diff
@@ -1,3 +1,77 @@
...
+  end
+
+  def destroy
+    @goal.destroy
+    load_goals_for_index
+    respond_to do |format|
+      format.turbo_stream { render :update }
+      format.html { redirect_to goals_path, status: :see_other }
+    end
+  end
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/controllers/goals_controller.rb:61]: Added in e09a5ba — `test/controllers/goals_controller_test.rb` mirrors `expenses_controller_test.rb`: auth, Current.user scoping, CRUD, validation errors, turbo_stream responses, and the no-funding-schedule setup pointer.

File: app/controllers/transaction_goals_controller.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+    GoalLinker.link(transaction: @transaction, goal: goal)
+    respond_with_drawer
+  end
+
+  def destroy
+    @transaction = Current.user.transactions.find(params.expect(:id))
+    GoalLinker.unlink(transaction: @transaction)
+    respond_with_drawer
+  end
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/controllers/transaction_goals_controller.rb:52]: Added in e09a5ba — `test/controllers/transaction_goals_controller_test.rb` mirrors the expense-linking controller test: successful link/unlink, non-positive (refund) rejection, under-funded-goal rejection, missing-param handling for turbo_stream vs html, RecordNotFound, and free-to-spend re-render.

File: app/services/goal_linker.rb [hunk 7]
```diff
@@ -0,0 +1,52 @@
+class GoalLinker
+  # Mirror of ExpenseLinker for goals. Walks the goal's untouched funded
+  # allocations oldest-first and consumes them up to `transaction.amount`.
+  # If the transaction is already linked to an expense, unlinks that first.
+  def self.link(transaction:, goal:)
+    Transaction.transaction do
+      ExpenseLinker.unlink(transaction: transaction) if transaction.expense_id.present?
+      unlink(transaction: transaction) if transaction.goal_id.present?
+
+      goal.with_lock do
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/services/goal_linker.rb:15]: Added in e09a5ba — `test/services/goal_linker_test.rb` mirrors `expense_linker_test.rb`: link/unlink, partial spend (oldest-first), overpay cap, `previous_due_on` restore through Jan 31 -> Feb clamping, skip-already-spent, and destroy-unlinks-first. Also added a cross-type test confirming linking a goal unlinks any prior expense link.

File: app/jobs/process_funding_events_job.rb [hunk 8]
```diff
@@ -22,10 +22,13 @@ def perform
...
-        .where(allocations: { funded_at: nil })
-        .distinct
-        .find_each do |user|
+    ids_via_expenses = User.joins(expenses: :allocations)
+                           .where(allocations: { funded_at: nil })
+                           .distinct.pluck(:id)
+    ids_via_goals = User.joins(goals: :allocations)
+                        .where(allocations: { funded_at: nil })
+                        .distinct.pluck(:id)
+    User.where(id: ids_via_expenses | ids_via_goals).find_each do |user|
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/jobs/process_funding_events_job.rb:31]: Acknowledged — leaving as-is for now. At current scale the two `pluck` + array union is negligible, and it keeps the "users with pending allocations" intent readable. Worth revisiting as an OR-over-subqueries if the user/allocation tables grow large.

**Comment by douglasrlee** [comment]
## Review feedback addressed (e09a5ba)

**Bugs**
- `goal-picker` now registered in the Stimulus manifest — the picker UI was inert without it.
- `transactions -> goals` FK now `on_delete: :nullify` — deleting a goal with a linked transaction no longer raises an FK violation.

**Migration safety**
- New `20260622053817_validate_goals_foreign_keys_and_constraints` validates the four new FKs + the `allocations_exactly_one_allocatable` CHECK constraint (matches the repo's `Validate*` pattern). Verified via rollback + re-migrate.
- Also fixed pre-existing rubocop offenses in the goals migration (string literals + non-reversible `remove_index`).

**Tests (60 added)** — mirroring the Expenses suite, restoring coverage:
- `GoalTest`, `GoalsControllerTest`, `GoalLinkerTest`, `TransactionGoalsControllerTest`, plus janedoe-scoped goal fixtures.

**Not changed**
- The `ProcessFundingEventsJob` pluck/union — left as-is (premature optimization at current scale); noted inline.

`bin/rails test` → 449 runs, 0 failures. `bin/rubocop` → clean.

**Comment by douglasrlee** [comment]
## Coverage restored to 100% (5f1001d)

The earlier codecov patch-coverage miss is closed. After the new tests, local SimpleCov reports **100.0% (1118/1118)** — up from the 99.55% the goal-allocation paths left behind.

Covered the 5 remaining lines:
- `app/models/allocation.rb:27` — `exactly_one_allocatable` error branch (tests for both-set and neither-set).
- `app/services/allocation_engine.rb:18–25` — the goals proposal loop in `propose_for` (per-goal proposal, mixed expense+goal schedule, skip fully-funded goal, plus `fund_pending_for` on a goal).

`CI=1 bin/rails test` → 455 runs, 0 failures.

File: app/models/allocation.rb [hunk 9]
```diff
@@ -17,4 +20,10 @@ def spent_amount_within_amount
 
     errors.add(:spent_amount, 'cannot exceed amount')
   end
+
+  def exactly_one_allocatable
+    return if expense_id.present? ^ goal_id.present?
+
+    errors.add(:base, 'must belong to either an expense or a goal, not both')
+  end
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/models/allocation.rb:28]: Fixed in ba08678 — message is now `must belong to exactly one of an expense or a goal`, which reads correctly for both the both-set and neither-set cases. Updated the AllocationTest assertions to match.

File: test/fixtures/allocations.yml [hunk 10]
```diff
@@ -9,3 +9,15 @@ netflix_second_pending:
   expense: netflix
   amount: 7.99
   funded_at:
+
+janes_goal_first:
+  funding_event: janes_paycheck_first
+  goal: janes_goal
+  amount: 8.00
+  funded_at: 2026-01-01T08:00:00Z
+
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply test/fixtures/allocations.yml:18]: Fixed in ba08678 — bumped `janes_goal_first.funded_at` to 2026-03-01 so it's on/after its event's `occurs_on`. (The mismatch came from moving `janes_paycheck_first` to March to dodge a date collision with johndoe's job-test events; the funded_at didn't follow.)

File: app/services/allocation_engine.rb [hunk 11]
```diff
@@ -55,26 +50,34 @@ def self.fund_pending_for(user)
...
+                                .order('expenses.due_on ASC, expenses.created_at ASC')
+                                .to_a
+
+    pending_goal = Allocation.eager_load(:goal)
+                             .where(goals: { user_id: user.id }, funded_at: nil)
+                             .order('goals.due_on ASC, goals.created_at ASC')
+                             .to_a
+
+    (pending_expense + pending_goal)
+      .sort_by { |a| [ a.expense&.due_on || a.goal&.due_on, a.created_at ] }
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/services/allocation_engine.rb]: Fixed in edd0e8d — the merged sort now keys on the expense/goal's `created_at` (`item = a.expense || a.goal; [item.due_on, item.created_at]`), matching each per-type SQL `order` and restoring the stable same-due-date funding priority. Added a regression test where the allocation `created_at` order is the inverse of the item `created_at` order (goal created first, but its allocation built second by `propose_for`): with a partial balance the earlier-created goal funds first, which would fail under the old allocation-keyed sort.

File: app/services/expense_linker.rb [hunk 12]
```diff
@@ -13,6 +13,7 @@ class ExpenseLinker
   # rollback of the prior expense) and then links to the new one.
   def self.link(transaction:, expense:)
     Transaction.transaction do
+      GoalLinker.unlink(transaction: transaction) if transaction.goal_id.present?
       unlink(transaction: transaction) if transaction.expense_id.present?
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/services/expense_linker.rb:22]: Fixed in 2495a6f — `link` now calls `transaction.lock!` at the top of the surrounding `Transaction.transaction` block, so concurrent expense/goal link requests serialize per-transaction (the transaction row is locked before either FK is read). Also added a DB-level `transactions_at_most_one_link` CHECK (`num_nonnulls(expense_id, goal_id) <= 1`) as a backstop.

File: app/services/goal_linker.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/services/goal_linker.rb:15]: Fixed in 2495a6f — `link` now calls `transaction.lock!` at the top of the surrounding `Transaction.transaction` block, so concurrent expense/goal link requests serialize per-transaction (the transaction row is locked before either FK is read). Also added a DB-level `transactions_at_most_one_link` CHECK (`num_nonnulls(expense_id, goal_id) <= 1`) as a backstop.

File: db/migrate/20260622053816_create_goals_and_extend_allocations_for_goals.rb [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+                         'num_nonnulls(expense_id, goal_id) = 1',
+                         name: 'allocations_exactly_one_allocatable',
+                         validate: false
+
+    add_column :transactions, :goal_id, :uuid
+    add_index :transactions, :goal_id, algorithm: :concurrently
+    # on_delete: :nullify so deleting a goal clears the link instead of raising
+    # an FK violation (mirrors the transactions/expenses fix in
+    # CascadeNullifyTransactionExpenseLinkingFks)
+    add_foreign_key :transactions, :goals, on_delete: :nullify, validate: false
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply db/migrate/20260622053816_create_goals_and_extend_allocations_for_goals.rb:56]: Added in 2495a6f — `transactions_at_most_one_link` CHECK (`num_nonnulls(expense_id, goal_id) <= 1`; `<= 1` because an unlinked transaction has neither set) added in the creation migration with `validate: false`, plus a `validate_check_constraint` in the follow-up `20260622053817` migration — mirroring the `allocations_exactly_one_allocatable` pattern. Paired it with `transaction.lock!` in both linkers to close the race at the source, and added a regression test asserting the DB rejects a both-linked state.

File: db/migrate/20260622053816_create_goals_and_extend_allocations_for_goals.rb [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    add_column :allocations, :goal_id, :uuid
+    add_index :allocations, :goal_id, algorithm: :concurrently
+
+    add_foreign_key :allocations, :goals, validate: false
+
+    # Replace the blanket unique index with two partial ones (one per type)
+    safety_assured do
+      remove_index :allocations, column: [ :funding_event_id, :expense_id ],
+                                 name: 'index_allocations_on_funding_event_id_and_expense_id'
+    end
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply db/migrate/20260622053816_create_goals_and_extend_allocations_for_goals.rb]: Fixed in ce1e9e4 — the `remove_index` now passes `algorithm: :concurrently` (safe since `disable_ddl_transaction!` is already set), so dropping the old blanket unique index uses `DROP INDEX CONCURRENTLY` instead of taking an ACCESS EXCLUSIVE lock. Dropping concurrently also satisfies strong_migrations, so I removed the now-unneeded `safety_assured` wrapper. Verified via rollback + re-migrate.

File: app/views/transactions/_transaction_detail.html.erb [hunk 15]
```diff
@@ -228,5 +228,90 @@
...
+          </div>
+        </div>
+      <% elsif transaction.expense.blank? %>
+        <% user_goals = Current.user.goals.order(:name).to_a %>
+        <div class="flex items-center justify-between gap-3">
+          <span class="text-sm text-base-content/50">Goal</span>
+          <button data-action="toggle#toggle" class="btn btn-ghost btn-sm text-primary">
+            Assign to goal
+          </button>
+        </div>
```
[the remark this answers is not in the document]
Comment (douglasrlee) [reply app/views/transactions/_transaction_detail.html.erb:262]: Done in ce1e9e4 — went with always showing both sections. When a transaction is linked to one, the other section now renders a disabled placeholder ("Unlink the expense to assign a goal" / "Unlink the goal to assign an expense") instead of vanishing, keeping the pickers mutually exclusive. Added controller tests asserting each placeholder renders.
